src/sources/socket/mod.rs

```diff
@@ -245,11 +268,103 @@ mod test {
         );
     }
 
+    #[test]
+    fn tcp_shutdown_simple() {
+        let source_name = "tcp_shutdown_simple";
+        let (tx, rx) = mpsc::channel(2);
+        let addr = next_addr();
+
+        let mut shutdown = ShutdownCoordinator::new();
+        let (shutdown_signal, _) = shutdown.register_source(source_name);
+
+        // Start TCP Source
+        let server = SocketConfig::from(TcpConfig::new(addr.into()))
+            .build(source_name, &GlobalOptions::default(), shutdown_signal, tx)
+            .unwrap();
+        let mut rt = runtime::Runtime::new().unwrap();
+        let source_handle = oneshot::spawn(server, &rt.executor());
+        wait_for_tcp(addr);
+
+        // Send data to Source.
+        rt.block_on(send_lines(addr, vec!["test".to_owned()].into_iter()))
+            .unwrap();
+
+        let event = rx.wait().next().unwrap().unwrap();
+        assert_eq!(
+            event.as_log()[&event::log_schema().message_key()],
+            "test".into()
+        );
+
+        // Now signal to the Source to shut down.
+        shutdown.shutdown_source_begin(source_name);
+        let deadline = Instant::now() + Duration::from_secs(10);
+        let shutdown_success =
+            shutdown.shutdown_source_end(&mut rt, source_name.to_string(), deadline);
+        assert_eq!(true, shutdown_success);
+
+        // Ensure source actually shut down successfully.
+        rt.block_on(source_handle).unwrap();
+    }
+
+    #[test]
+    fn tcp_shutdown_infinite_stream() {
+        // It's important that the buffer be large enough that the TCP source doesn't have
```

Comment:
If anyone's curious what I spent most of the last day investigating, my first version of this test used a buffer smaller than the amount of events we were sending/consuming (intentionally).  The problem is that the TCP source calls `.forward()` on its input stream and forwards that into the output channel.  If the output channel is full the Source Task will go to sleep until the channel has openings again, which in this test it never will.  Since the `.forward` is being called on a stream that has already been combined with the ShutdownSignal Tripwire via a call to `take_until` it'd be nice if notifying that Tripwire would wake up the Task, but unfortunately that doesn't seem to be the case.  I'm not sure if there's any way to get around that.  Maybe writing our own version of `Forward` that took an 'until' future and checked before ever trying to send into its `Sink`, but if Tasks are always woken up at the lowest level and only by the thing that caused them to return NotReady changing, then I'm not sure that'll help.

Thankfully, in real systems this is less likely to be an issue as usually you'll have a sink always pulling from the output buffer.  Although it does mean that if a Sink got backed up/blocked then graceful shutdown of its input source would fail and require force shutdown.  But that may just be one of the intended use cases for force shutdown.

It also means that when we update `RunningTopology::stop()` to use the `ShutdownCoordinator` we'll need to make sure to fully shut down all sources before starting shutdown of the Sinks, although this is likely something we'd want to do anyway to minimize dropped Events.

Reply:
> It also means that when we update `RunningTopology::stop()` to use the `ShutdownCoordinator` we'll need to make sure to fully shut down all sources before starting shutdown of the Sinks, although this is likely something we'd want to do anyway to minimize dropped Events.

As much as possible, it'd be nice if shutdown meant shutting down sources. They would finish anything in flight and close their streams, which would naturally cascade through the rest of the pipelines. In that approach, sinks and transform don't need to do anything special beyond see that their input stream has closed.

Reply:
The problem is that if sending to the Sink (via try_start_send) returns NotReady the whole Task goes to sleep and only gets woken up if the Sink becomes available again.  So the Forward future won't figure out that the input Stream has been closed.  I don't think there's anything that can be done about that, and the behavior makes sense, it just introduces this tricky edge case in shutdown

Reply:
It should always hit the `stream.poll` on the match before ever hitting try_start_send so in theory the "Async::Ready(None)" from the stream should still wake up the task right?

Reply:
Hmm... that wasn't my experience when testing.  I'm still a little hazy on the details of when exactly Tasks get parked() and unparked() and where in the future chain they resume when unparked, but everything I read seemed to suggest that when you try to do something to an external system and get back NotReady you go to sleep and register your task to be resumed when the system you blocked on says it is now available.  That would suggest that when Forward tries to send into its Sink (which is actually a mpsc::Sender) and finds that the channel is full, it will go to sleep and only be woken once the Sink/channel has capacity to receive new inputs again.  So if nothing ever reads from the Receiver end of the channel ever again, the Forward future won't ever notice that its input stream has closed.